### DISCUSSION TITLE
DropdownMenu.Item: Handle async functionality onClick

### DISCUSSION THREAD

**Discussion by ForrestCogna** [formaat-design/reshaped#220] [Q&A] [answered]
How would you create an asynchronous menu item? I'm having trouble handling the active state of the dropdown menu:
[Codesandbox example](https://codesandbox.io/p/sandbox/async-dropdown-menu-368zj4?file=%2Fsrc%2FApp.tsx%3A3%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clscy786300063b6ky3wjizoa%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clscy786200023b6ka4h62h6g%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clscy786200033b6k0mneg6il%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clscy786200053b6kn0xlbi82%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clscy786200023b6ka4h62h6g%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clscy786200013b6ks1fsab9k%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpublic%252Findex.html%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clscyiozp00023b6krdldbvnf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A3%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A3%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clscy786200023b6ka4h62h6g%2522%252C%2522activeTabId%2522%253A%2522clscyiozp00023b6krdldbvnf%2522%257D%252C%2522clscy786200053b6kn0xlbi82%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clscy786200043b6kosy7dqvx%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clscy786200053b6kn0xlbi82%2522%252C%2522activeTabId%2522%253A%2522clscy786200043b6kosy7dqvx%2522%257D%252C%2522clscy786200033b6k0mneg6il%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clscy786200033b6k0mneg6il%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
```ts
const MyAsyncDropdown = () => {
  const [pending, setPending] = useState(false)
  return (
     <DropdownMenu>
      <DropdownMenu.Trigger>
        {(attributes) => <Button attributes={attributes}>Open Menu</Button>}
      </DropdownMenu.Trigger>
      <DropdownMenu.Content>
        <DropdownMenu.Item
          icon={pending ? <Loader /> : undefined}
          onClick={async () => {
            setPending(true);
            await new Promise((resolve) => setTimeout(resolve, 2000));
            setPending(false);
            // only close menu after promise is resolved
            // if error (not shown here) do not close, but toggle error specific props in menu.item
          }}
        >
          Download file
        </DropdownMenu.Item>
      </DropdownMenu.Content>
    </DropdownMenu>
  )
}

**Comment by blvdmitry** [comment, marked as answer]
Hey, you can do this with the controlled component behavior. DropdownMenu supports an `active` flag which you can change after the promise is resolved based on your own `useState`

**Reply by ForrestCogna** [reply]
Thanks!

**Comment by ForrestCogna** [comment]
Follow up question: when using the controlled version of a DropdownMenu, how can the onClickOutside behavior be handled? I thought it would be to use the onClose prop, however this is also fired on menu item click so that is not a valid solution in the case of this discussion

**Reply by blvdmitry** [reply]
Right now it doesn't differentiate between the two in terms of handling but we can implement it for the next release in case you're looking for a way to handle that. There are possible workarounds potentially if you check `document.activeElement` in the handle to see if an element inside the popover content is focused or something else, but definitely easier with the native support for this feature
